core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -159,6 +159,15 @@ private[spark] class TaskSchedulerImpl(
     this.synchronized {
       val manager = createTaskSetManager(taskSet, maxTaskFailures)
       activeTaskSets(taskSet.id) = manager
+      val attemptsByStage = activeTaskSets.values.map(_.taskSet).groupBy(_.stageId)
```

Comment:
> we shouldn't even have two task sets for the same stage

Yeah, that's the other thing that was bugging me. This looks too late - why is the task even being submitted? I assume that everything gets cleaned up properly when the exception is thrown (which was the reason I asked the original question).